src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -0,0 +1,142 @@
+        if (auto multi_subgraph_op = dynamic_pointer_cast<op::util::MultiSubGraphOp>(op)) {
+            for (size_t i = 0; i < multi_subgraph_op->get_internal_subgraphs_size(); i++) {
+                if (auto sub_graph = multi_subgraph_op->get_function(i))
+                    rewritten |= shared_node_optimization(sub_graph, rules);
+            }
+        }
+        for (auto& output : op->outputs()) {
+            const auto& target_inputs = output.get_target_inputs();
+            if (target_inputs.size() <= 1)
+                continue;  // nothing to optimize
+            unordered_map<Node::type_info_t, vector<Node*>> type_to_node;
+            for (const auto& input : target_inputs) {
+                auto node = input.get_node();
+                if (node && rules.count(node->get_type_info()))
+                    type_to_node[node->get_type_info()].push_back(node);
+            }
+            for (auto& item : type_to_node) {
+                const auto& shared_nodes = item.second;
+                if (shared_nodes.size() < 2)
+                    continue;
+                const auto& ops_type = item.first;
+                const auto& are_equal = rules.at(ops_type);
+
+                std::vector<bool> visited_nodes(shared_nodes.size(), false);
+                for (size_t i = 0; i < visited_nodes.size(); ++i) {
+                    if (visited_nodes[i])
+                        continue;
+                    const auto& root_op = shared_nodes[i];
+                    visited_nodes[i] = true;
+                    for (size_t j = i + 1; j < visited_nodes.size(); ++j) {
+                        if (visited_nodes[j])
+                            continue;
+                        const auto& child_op = shared_nodes[j];
+                        if (are_equal(root_op, child_op)) {
+                            rewritten |= replace_output_update_name(child_op->output(0), root_op->output(0));
+                            visited_nodes[j] = true;
+                        }
+                    }
+                }
+            }
+        }
+    }
+    return rewritten;
+}
+
+bool inputs_from_same_source_or_equal_constants(const Node* lhs, const Node* rhs) {
+    if (lhs->get_input_size() != rhs->get_input_size())
+        return false;
+    size_t input_size = lhs->get_input_size();
+    for (size_t i = 0; i < input_size; ++i) {
+        if (lhs->input_value(i) == rhs->input_value(i))
+            continue;
+        auto lhs_constant = as_type_ptr<v0::Constant>(lhs->get_input_node_shared_ptr(i));
+        auto rhs_constant = as_type_ptr<v0::Constant>(rhs->get_input_node_shared_ptr(i));
+        if (!lhs_constant || !rhs_constant)
+            return false;
+        if (lhs_constant->get_element_type() != rhs_constant->get_element_type())
+            return false;
+        const auto& lhs_shape = lhs_constant->get_shape();
+        if (lhs_shape != rhs_constant->get_shape() || shape_size(lhs_shape) > 10)
```

Comment:
why > 10?
could you add comment here?

Reply:
I can add the comment later -- right now green pre-commit is rare